File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> m(256,-1);

        int l=0,r=0;
        int n = s.size();
        int cnt=0;

        while(r<n){
            if(m[s[r]] != -1){
                l = max(m[s[r]] + 1, l);
            }

            m[s[r]] = r;

            cnt = max(cnt, r-l+1);
            r++;
        }
        return cnt;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        st = []

        lt = [-1] * n
        for i in range(n):
            while st and heights[st[-1]] >= heights[i]:
                st.pop()
            if st:
                lt[i] = st[-1]
            st.append(i)

        st = []
        rt = [n] * n
        for i in range(n-1,-1,-1):
            while st and heights[st[-1]] >= heights[i]:
                st.pop()
            if st:
                rt[i] = st[-1]
            st.append(i)
        
        mxa = 0
        for i in range(n):
            lt[i] += 1
            rt[i] -= 1
            mxa = max(mxa, heights[i] * (rt[i] - lt[i] + 1))
        
        return mxa
        